src/algorithm/binarySearchTree.ts

```diff
@@ -0,0 +1,171 @@
+	}
+
+	private recursiveSet(node: TreeNode<Key, Val> | null, key: Key, val: Val): TreeNode<Key, Val> {
+		if (node == null) {
+			return new TreeNode(key, val)
+		}
+
+		if (key < node.key) {
+			node.left = this.recursiveSet(node.left, key, val)
+		} else if (key > node.key) {
+			node.right = this.recursiveSet(node.right, key, val)
+		} else {
+			node.val = val
+		}
+
+		return node
+	}
+
+	public get(key: Key): Val | null {
+		return this.recursiveGet(this.root, key)
+	}
+
+	private recursiveGet(node: TreeNode<Key, Val> | null, key: Key): Val | null {
+		if (node == null) return null
+		if (key < node.key) {
+			return this.recursiveGet(node.left, key)
+		} else if (key > node.key) {
+			return this.recursiveGet(node.right, key)
+		} else {
+			return node.val
+		}
+	}
+
+	private recursiveDelMin(node: TreeNode<Key, Val>): TreeNode<Key, Val> {
+		if (node.left == null) {
+			return node.right
+		}
+
+		node.left = this.recursiveDelMin(node.left)
+		return node
+	}
+
+	public delete(key: Key) {
+		if (this.root) {
+			this.root = this.delRecursively(this.root, key)
+		}
+	}
+
+	public clear() {
+		this.root = null
+	}
+
+	private delRecursively(node: TreeNode<Key, Val>, key: Key): TreeNode<Key, Val> {
+		if (key < node.key) {
+			node.left = this.delRecursively(node.left, key)
+		} else if (key > node.key) {
+			node.right = this.delRecursively(node.right, key)
+		} else {
+			if (node.right == null) {
+				return node.left
+			}
+			if (node.left == null) {
+				return node.right
+			}
+
+			var t = node
+			node = this.min(node.right)
+			node.right = this.recursiveDelMin(t.right)
+			node.left = t.left
+		}
+		return node
+	}
+
+	private min(node: TreeNode<Key, Val>): TreeNode<Key, Val> {
+		if (node.left == null) {
+			return node
+		}
+
+		return this.min(node.left)
+	}
+
+	public entries(): TreeNode<Key, Val>[] {
+		var array: TreeNode<Key, Val>[] = []
+		this.entriesRecursively(this.root, array)
+		return array
+	}
+
+	private entriesRecursively(node: TreeNode<Key, Val> | null, array: TreeNode<Key, Val>[]) {
+		if (node == null) {
+			return
+		}
+
+		this.entriesRecursively(node.left, array)
+		array.push(node)
+		this.entriesRecursively(node.right, array)
+	}
+
+	public keys(): Key[] {
+		var array: Key[] = []
+		this.keyRecursively(this.root, array)
+		return array
+	}
+
+	public has(key: Key): boolean {
+		var keys = this.keys()
+		return keys.indexOf(key) > -1
+	}
+
+	private keyRecursively(node: TreeNode<Key, Val> | null, array: Key[]) {
+		if (node == null) {
+			return
+		}
+
+		this.keyRecursively(node.left, array)
+		array.push(node.key)
+		this.keyRecursively(node.right, array)
+	}
+
+	public values(): Val[] {
+		var array: Val[] = []
+		this.valRecursively(this.root, array)
+		return array
+	}
+
+	private valRecursively(node: TreeNode<Key, Val> | null, array: Val[]) {
+		if (node == null) {
+			return
+		}
+
+		this.valRecursively(node.left, array)
+		array.push(node.val)
+		this.valRecursively(node.right, array)
```

Comment:
hum is there a reason why you wrote `keyRecursively` and `valRecursively`, they really look like a copy pasta of `entriesRecursively` 😄

Is it not possible to simply reuse `entriesRecursively` and a `.map` to get the value you need ?
```ts
	public keys(): Key[] {
		var array: Val[] = []
		this.entriesRecursively(this.root, array).map(x => x.Key)
		return array
	}

	public values(): Val[] {
		var array: Val[] = []
		this.entriesRecursively(this.root, array).map(x => x.Val)
		return array
	}
```
I didnt check but it should kinda works

Reply:
kk I thought If I call `entriesRecursively ` and filter them for keys and values, I have to iterate though them again.
So The complex of these operation would become `2LogN` but at the end of the day it would be O(logN)  
So It should be fine!

Reply:
those two function are design to get all keys or all values of a tree, exporting data to reuse them somewhere else without the tree structure.

This kind of operation is slow both CPU and memory wise, so I'm fine if those two are slightly slower, there are more here for compatibility and ease of use than performance 😄